### PR TITLE
feat(node): release from CI with npm provenance statements

### PR DESCRIPTION
**Pull request by gmeligio**
This PR adds `npmProvenance` to allow increasing supply-chain security for a package.

npm and pnpm support it, but Yarn still needs to, as per [npm guide](https://docs.npmjs.com/generating-provenance-statements). It's supported only for public packages.

It will be false by default.

## References

https://docs.npmjs.com/generating-provenance-statements
https://docs.npmjs.com/cli/v10/using-npm/config#provenance

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mrgrain** [comment]
This is awesome! Now projen's default release workflow uses `publib` which interally uses `npm`, so I don't think we need to forcefully limit this for yarn users either. Maybe just write this in the docs?

**Comment by mrgrain** [comment]
PS: The failing build should fix once this PR is merged: https://github.com/projen/projen/pull/3331

**Comment by gmeligio** [comment]
Thank you, @mrgrain, for fixing the build.

I have removed the Yarn limitation, leaving `provenance`=true  by default for public packages. Node package managers can indeed be mixed in different development stages. I added documentation as you suggested.

As a consequence, `projen` itself will have provenance enabled. I think it's a good default.

**Comment by mrgrain** [comment]
We will also need to add the extra `id-token: write` permission to the release workflow:

https://docs.npmjs.com/generating-provenance-statements#example-github-actions-workflow

**Comment by mrgrain** [comment]
I also have mild concerns about compatibility. How about we don't turn it on by default just yet. Enable it in projen and I can run it on a bunch of repos I have access to. After a month or so we can make it the new default.

**Comment by gmeligio** [comment]
Thanks for reviewing it so promptly.

Sounds good. I changed the default to false. False is actually npm's default. But I think it won't cause compatibility issues.

As you said, it would be nice to test in real repositories and later change the default to true.

**Comment by mrgrain** [comment]
> Sounds good. I changed the default to false. False is actually npm's default. But I think it won't cause compatibility issues.

I think this is one of my concerns: https://docs.npmjs.com/generating-provenance-statements#example-gitlab-ci-job
Once we enable it by default, presumingly npm would try to sign the release but it would fail unless the user change their gitlab config. This is less of an issue for GitHub because we somewhat control the release workflow. But there's currently not GitLab equivalent.

This made me think a bit. Maybe the main option for this should to be on the [Release](https://projen.io/docs/api/release#releaseoptions-) component. We could then use the environment variable instead of adding it package.json (might just be easier). My next leap is than that other ecosystems can sign releases as well! Most notably [PyPi](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/)... tl;dr the best place for this flag is probably in [NpmPublishOptions](https://projen.io/docs/api/release#npmpublishoptions-) and that's where we can change the default eventually.

Of course we can still keep the option on the NodePackage, just need to document it well. And potentially be clever with defaults.

**Comment by gmeligio** [comment]
You were right. It makes more sense to persist the provenance configuration in the CI workflow than in the package because it needs other features from the CI provider, such as permissions.

Your thoughts helped me to realize how to change the implementation. I used a similar approach to how `npmTokenSecret` is handled between `NodePackage and `Release` by setting `npmProvenance` in `NodePackage` and passing it from there to `Release.publishToNpm`.

I tested it in a project I use on GitHub and it worked correctly.

I left it for Github for now, which is better integrated with Projen components. I use more Gitlab in general, but to do this provenance part, I think it should wait until after https://github.com/projen/projen/issues/2438 is solved. 

Regarding other ecosystems, it would be nice to have it in PyPi and others. Both npm and PyPi _signatures_ are new to me. I'd like to dive into PyPi's details later and see how it can be done.

**Comment by mrgrain** [comment]
Thanks, I'll review again soon.

If you're interested, for Python the equivalent would be [PyPI Trusted Publishers](https://docs.pypi.org/trusted-publishers/).

File: src/javascript/node-package.ts [hunk 1]
```diff
@@ -494,6 +509,11 @@ export class NodePackage extends Component {
    */
   public readonly npmAccess: NpmAccess;
 
+  /**
+   * Wether provenance statements should be generated when package is published.
```
Comment (mrgrain) [inline src/javascript/node-package.ts:513]:
```suggestion
   * Wether provenance statements should be generated when package is published.
   * @default false
```

**Comment by mrgrain** [comment]
- Made some tweaks to docs
- Changed my mind about the default and it's now enabled by default

**Comment by mrgrain** [comment]
@gmeligio This is ready to go now, but I'd like to give you chance to look at my changes first. Thank you so much for this PR though, this is huge!

**Comment by gmeligio** [comment]
Thank you for all the changes. It reads much better now.  It looks great and ready to go. I just did a minor update to a doc comment.

**Review by mrgrain — APPROVED** [review]

**Comment by gmeligio** [comment]
I made a mistake in this PR and created the fix in https://github.com/projen/projen/pull/3378 . I'm sorry for that. Could you review @mrgrain please when you have time?

**Pull request merged**